Library_Engine/Query/Libraries.cs

```diff
@@ -74,47 +92,94 @@ public static void RefreshLibraries()
+            Dataset dataset;
+
+            if (!m_datasets.TryGetValue(name, out dataset))
             {
-                //string libraryString = LibraryString(name);
+
+                m_deserialisationEvents[name] = new List<Tuple<oM.Reflection.Debugging.EventType, string>>();
 
                 string[] entries = LibraryString(name);
 
-                if (entries == null)
-                    return new List<IBHoMObject>();
+                if (entries == null || entries.Length == 0)
+                {
+                    AddDeserialisationEvent(name, oM.Reflection.Debugging.EventType.Warning, "No dataset with the name " + name + " could be found.");
+                    return new Dataset();
+                }
+                else if (entries.Length == 1)
+                {
 
-                //string[] entries = libraryString.Split('\n');
-                objects = new List<IBHoMObject>();
+                    IBHoMObject obj;
+                    try
+                    {
+                        obj = BH.Engine.Serialiser.Convert.FromJson(entries[0]) as IBHoMObject;
+                    }
+                    catch
+                    {
+                        AddDeserialisationEvent(name, oM.Reflection.Debugging.EventType.Error, "Failed to deserialise the dataset with name " + name);
+                        return new Dataset();
+                    }
+
+                    dataset = obj as Dataset;
 
-                //Parse the library from json to IBHoMObjects
-                foreach (string entry in entries)
+                    if (dataset == null)
+                    {
+                        dataset = new Dataset() { Data = new List<IBHoMObject> { obj } };
+                        AddDeserialisationEvent(name, oM.Reflection.Debugging.EventType.Warning, "No Source information is available for the dataset named " + name);
+                    }
+                }
+                else
                 {
-                    if (!string.IsNullOrWhiteSpace(entry))
+
+                    List<IBHoMObject> objects = new List<IBHoMObject>();
+
+                    //Parse the library from json to IBHoMObjects
+                    foreach (string entry in entries)
                     {
-                        IBHoMObject obj;
-                        try
+                        if (!string.IsNullOrWhiteSpace(entry))
                         {
-                            obj = (IBHoMObject)BH.Engine.Serialiser.Convert.FromJson(entry);
+                            IBHoMObject obj;
+                            try
+                            {
+                                obj = (IBHoMObject)BH.Engine.Serialiser.Convert.FromJson(entry);
+                            }
+                            catch
+                            {
+                                AddDeserialisationEvent(name, oM.Reflection.Debugging.EventType.Warning, "Failed to deserialise at least one item in the dataset named " + name);
```

Comment:
Would it be helpful to state which `entry` has failed to deserialise?

Reply:
Presume it could be useful, but two points:
1. As it failed to deserialise, I will not really have anything to look at for a generic case to state what it was. As in, failing to deserialise means I dont have a BHoMObject to take name or similar from to display. Could paste the whole string out or something like that, but not sure how useful that would be?
1. In general when we raise multiples of the same error, it generally feels as it is more clogging the UI than helping out + at least GH has a maximum error count of 15.

Please let me know if you are ok with this, or want me to update!

Reply:
I'm happy with that, it makes sense to me, happy to go ahead with merging without changes :smile: